ledger-core/v2/virtual/handlers/vdelegatedrequestfinished.go

```diff
@@ -47,6 +49,32 @@ func (s *SMVDelegatedRequestFinished) GetStateMachineDeclaration() smachine.Stat
 }
 
 func (s *SMVDelegatedRequestFinished) Init(ctx smachine.InitializationContext) smachine.StateUpdate {
+	return ctx.Jump(s.stepGetObject)
+}
+
+func (s *SMVDelegatedRequestFinished) stepGetObject(ctx smachine.ExecutionContext) smachine.StateUpdate {
+	s.objectSharedState = s.objectCatalog.GetOrCreate(ctx, s.Payload.Callee)
+
+	var semaphoreReadyToWork smachine.SyncLink
+
+	action := func(state *object.SharedState) {
+		semaphoreReadyToWork = state.ReadyToWork
+	}
+
+	switch s.objectSharedState.Prepare(action).TryUse(ctx).GetDecision() {
+	case smachine.NotPassed:
+		return ctx.WaitShared(s.objectSharedState.SharedDataLink).ThenRepeat()
+	case smachine.Impossible:
+		ctx.Log().Fatal("failed to get object state: already dead")
+	case smachine.Passed:
+	default:
+		panic(throw.NotImplemented())
+	}
+
+	if ctx.AcquireForThisStep(semaphoreReadyToWork).IsNotPassed() {
+		return ctx.Sleep().ThenRepeat()
```

Comment:
hmm, not sure who would wake up sm in this case. We have the same logic in execute @bigbes @cyraxred is this a proper behavior? I'm a bit confused

Reply:
In short: SyncLinks have embedded mechanizm to wake up waiters. Was answered in a chat in details.